File: Algoritmos/N Rainhas/NRainhas.cpp
#include <iostream>
#include <string>
#include <sstream>

using namespace std;

#define RAINHA 1
#define VAZIO 0

class NRainhas
{
	// Atributos
private:
	int **tabuleiro;
	int dimensao;

private:
	// Retorna true caso haja algum conflito na solução
	bool conflito(int linha, int coluna)
	{
		for(int i = 0; i < linha; i++)
		{
			if(tabuleiro[i][coluna] == RAINHA)
			{
				return false;
			}
		}

		for(int lin = 0; lin < linha; lin++)
		{
			for(int col = 0; col < this->dimensao; col++)
			{
				if((tabuleiro[lin][col] == RAINHA) &&
					((linha - lin) == (coluna - col) ||
					(linha - lin) == (col - coluna)))
				{
					return false;
				}
			}
		}

		return true;
	}

public:
	// Construtor
	NRainhas(int dimensao)
	{
		this->dimensao = dimensao;
		this->tabuleiro = new int*[dimensao];

		for(int i = 0; i < this->dimensao; i++)
		{
			tabuleiro[i] = new int[dimensao]();
		}
	}

	~NRainhas()
	{
		if(tabuleiro)
		{
			for(int i = 0; i < this->dimensao; i++)
			{
			
				if(tabuleiro[i])
				{
					delete tabuleiro[i];
					tabuleiro[i] = nullptr;
				}
		
			}

			delete[] tabuleiro;
			tabuleiro = nullptr;
		}
	}

	// backtracking
	void encontrarSolucoes(int linha)
	{
		if(linha == this->dimensao)
		{
			print();
		}
		else
		{
			for(int coluna = 0; coluna < this->dimensao; coluna++)
			{
				if(tabuleiro[linha][coluna] == VAZIO)
				{
					tabuleiro[linha][coluna] = RAINHA;

					if(conflito(linha, coluna))
					{
						encontrarSolucoes(linha + 1);
					}

					tabuleiro[linha][coluna] = VAZIO;
				}
			}
		}
	}

	void print()
	{
		for(int linha = 0; linha < this->dimensao; linha++)
		{
			string resultado;			

			for(int coluna = 0; coluna < this->dimensao; coluna++)
			{
				ostringstream ss;
				ss << tabuleiro[linha][coluna];
				resultado += ss.str() + " ";
			}

			cout << resultado << endl;
		}

		cout << endl;
	}
};


int main(int argc, char* argv[])
{
	NRainhas jogo(4);
	jogo.encontrarSolucoes(0);
	return 0;
}

File: Algoritmos/N Rainhas/NRainhas.py
class NRainhas:
    def __init__(self, dimensao):
        self.tabuleiro = [[0 for x in range(dimensao)] for y in range(dimensao)]
        self.RAINHA = 1
        self.VAZIO = 0

    def __str__(self):
        resultado = ""

        for linha in range(0, len(self.tabuleiro)):
            resultado += str(self.tabuleiro[linha])
            resultado += "\n"
            
        return resultado

    
    def temConflito(self, linha, coluna):
        for i in range(0, linha):
            if self.tabuleiro[i][coluna] == self.RAINHA:
                return True

        for lin in range(0, linha):
            for col in range(0, len(self.tabuleiro[linha])):
                if (self.tabuleiro[lin][col] == self.RAINHA) and ((linha - lin) == (coluna - col) or (linha - lin) == (col - coluna)):
                    return True

        return False


    def encontrarSolucoes(self, linha):
        if linha == len(self.tabuleiro):
            print(self)
        else:
            for coluna in range(0, len(self.tabuleiro[linha])):
                if(self.tabuleiro[linha][coluna] == self.VAZIO):
                    self.tabuleiro[linha][coluna] = self.RAINHA

                    if self.temConflito(linha, coluna) == False:
                        self.encontrarSolucoes(linha + 1)

                    self.tabuleiro[linha][coluna] = self.VAZIO


  
jogo = NRainhas(4)
jogo.encontrarSolucoes(0)
